### PR TITLE
Implemented {n} notation for Collection::extract()

### PR DESCRIPTION
**Pull request by lorenzo**

### CODE REVIEW

**Comment by jrbasso** [comment]
:+1:

**Comment by ADmad** [comment]
Me every time trying to understand a diff relating to collections:

Line 1:  Okay, I _think_ I get this..
Line 2:  Hmm not sure, let me just skip to the next line...
Line 3:  Help! I am lost, get me out of here!!

**Comment by lorenzo** [comment]
@ADmad I think I can explain. What is the most confusing part for you?

**Comment by ADmad** [comment]
@lorenzo It's my laziness to get a better understanding of iterators in general. Hopefully I will remedy that someday :)

File: src/Collection/CollectionTrait.php [hunk 1]
```diff
@@ -158,11 +158,20 @@ public function reduce(callable $c, $zero = null)
...
      * {@inheritDoc}
      *
-     * @return \Cake\Collection\Iterator\ExtractIterator
      */
     public function extract($matcher)
     {
-        return new ExtractIterator($this->unwrap(), $matcher);
+        $extractor = new ExtractIterator($this->unwrap(), $matcher);
+
+        if (is_string($matcher) && strpos($matcher, '{n}') !== false) {
```
Comment (markstory) [inline src/Collection/CollectionTrait.php]: I guess in the future we could add `{*}` and `{s}`.

File: src/Collection/ExtractTrait.php [hunk 2]
```diff
@@ -53,6 +61,45 @@ protected function _propertyExtractor($callback)
...
+            }
+
+            if ($collectionTransform &&
+                !($data instanceof \Traversable || is_array($data))) {
+                return null;
+            }
+
+            if ($collectionTransform) {
+                $rest = implode('.', array_slice($path, $i));
+                return (new Collection($data))->extract($rest);
```
Comment (markstory) [inline src/Collection/ExtractTrait.php]: This magic isn't free.

File: src/Collection/CollectionTrait.php (hunk 1 above)
Comment (markstory) [inline src/Collection/CollectionTrait.php]: Given that it matches any key we should probably use `{*}`.

**Comment by lorenzo** [comment]
@markstory renamed `{n}` to `{*}`

**Comment by markstory** [comment]
Looks good to me :+1:

**Comment by markstory** [comment]
Needs docs @lorenzo :smile:

**Pull request merged**